specs/schemas/schema-schema.ipldsch

```diff
@@ -336,14 +341,30 @@ type TypeLink struct {
 ## Unions can be defined as representing in several different ways: see
 ## the documentation on the UnionRepresentation type for details.
 ##
-## The set of types which the union can contain are specified in a map
-## inside the representation field.  (The key type of the map varies per
-## representation strategy, so it's not possible to keep on this type directly.)
-##
 type TypeUnion struct {
+	members [UnionMember]
 	representation UnionRepresentation
 }
 
+## UnionMember is a type for identifying the members of a union.
+## Most commonly, this is simply a TypeName string;
+## however, it can also be a UnionMemberInlineDefn,
+## which is used to allow inline link definitions within a kinded union as a shorthand
+## (rather than requiring all links be declared as a named type before being usable within a union).
+type UnionMember union {
+	| TypeName string
+	| UnionMemberInlineDefn map
+} representation kinded
+
+## UnionMemberInlineDefn is a very similar purpose to InlineDefn,
+## but found specifically within UnionMember.
+## It only allows describing a link type (and not maps nor lists, as InlineDefn does),
+## which is a constraint applied to union membership largely to make sure
+## if there are errors in processing unions, we can make legible messages about it!
+type UnionMemberInlineDefn union {
+	| TypeLink "link"
```

Comment:
Justification for limiting this to links and not inline maps and lists too?

Reply:
I think, organisationally and from a documentation perspective, inline definitions for maps and lists is probably not a good idea anyway, but is there reasoning beyond that which prevents this from being technically possible? Maybe document further why it's only links because if it's simply _undesireable_ then we could have requests to broaden this from folks who don't see that.

Reply:
Weak, actually.  We _could_ drop this type and use `InlineDefn` as one of the members of `UnionMember`.

Mostly, I just find it kind of scary to think about how I'd format error messages if someone wrote a type like this:

```
type Wow union {
	| {Oof:{Wat:nullable Wot}} map
} representation kinded
```

and it doesn't seem nice to read it; and it also seems like making the DSL parsers deal with that would be headachey.

Maybe it's less bad than I think, though.

Reply:
yeah, I feel the same gross-out at people doing those kinds of things, _but_ we allow them elsewhere so if we're excluding here then state a reason or it'll creep in in a future date by someone else's hand because they notice the inconsistency and don't see a reason for it

reason could just be something like "legibility and coherence breaks down once we start doing this inside unions, more so than in structs and other places `InlineDefn` is used"